### PR TITLE
smt2_dec now uses run() instead of system()

### PR DESCRIPTION
**Pull request by kroening**

### CODE REVIEW

File: src/solvers/smt2/smt2_dec.cpp [hunk 1]
```diff
@@ -74,88 +78,94 @@ decision_proceduret::resultt smt2_dect::dec_solve()
...
   {
   case solvert::BOOLECTOR:
-    command = "boolector --smt2 "
-            + smt2_temp_file.temp_out_filename
-            + " -m > "
-            + smt2_temp_file.temp_result_filename;
+    argv.push_back("boolector");
+    argv.push_back("--smt2");
+    argv.push_back(smt2_temp_file.temp_problem_filename);
+    argv.push_back("-m");
```
Comment (tautschnig) [inline src/solvers/smt2/smt2_dec.cpp]: I think `argv = { "boolector", "--smt2", smt2_temp_file.temp_problem_filename, "-m" };` might work with C++11?

File: src/solvers/smt2/smt2_dec.cpp [hunk 2]
```diff
@@ -44,22 +46,24 @@ std::string smt2_dect::decision_procedure_text() const
 
 smt2_temp_filet::smt2_temp_filet()
 {
-  temp_out_filename=get_temporary_file("smt2_dec_out_", "");
+  temp_problem_filename = get_temporary_file("smt2_dec_problem_", "");
```
Comment (tautschnig) [inline src/solvers/smt2/smt2_dec.cpp:49]: Could this cleanup (more on this below) please be put in a separate commit? It's surely a very good idea, but it renders (the rather important) git history useless.

File: src/solvers/smt2/smt2_dec.cpp [hunk 3]
```diff
@@ -44,22 +46,24 @@ std::string smt2_dect::decision_procedure_text() const
...
+  if(temp_problem_filename != "")
+    unlink(temp_problem_filename.c_str());
+
+  if(temp_stdout_filename != "")
+    unlink(temp_stdout_filename.c_str());
 
-  if(temp_result_filename!="")
-    unlink(temp_result_filename.c_str());
+  if(temp_stderr_filename != "")
+    unlink(temp_stderr_filename.c_str());
```
Comment (tautschnig) [inline src/solvers/smt2/smt2_dec.cpp]: Can we actually get rid of `smt2_temp_filet` completely? Back when this code was first written we probably hadn't yet got `temporary_filet` - we now do, and that should be the only place managing temporary files. If for some reason this is infeasible then surely `temp_dirt` would work? But all of this should be separate commit not to distract us from the more trivial `system` -> `run` transition.

File: src/solvers/smt2/smt2_dec.cpp (hunk 1 above)
Comment (kroening) [reply src/solvers/smt2/smt2_dec.cpp]: Ok, done

File: src/solvers/smt2/smt2_dec.cpp (hunk 2 above)
Comment (kroening) [reply src/solvers/smt2/smt2_dec.cpp:49]: Ok, split

File: src/solvers/smt2/smt2_dec.cpp (hunk 3 above)
Comment (kroening) [reply src/solvers/smt2/smt2_dec.cpp]: Yes, I did contemplate this; will look into it.

**Review by tautschnig — APPROVED** [review]
Ok, but it would be really nice to see the temp file business cleaned up. There is a lot of churn around these in the PR right now, and cleaning them up upfront might simplify this a lot.

File: src/solvers/smt2/smt2_dec.cpp [hunk 4]
```diff
@@ -44,22 +46,24 @@ std::string smt2_dect::decision_procedure_text() const
...
 }
 
 smt2_temp_filet::~smt2_temp_filet()
 {
-  temp_out.close();
+  temp_problem.close();
 
-  if(temp_out_filename!="")
-    unlink(temp_out_filename.c_str());
+  if(temp_problem_filename != "")
```
Comment (peterschrammel) [inline src/solvers/smt2/smt2_dec.cpp]: `! .empty()` (several occurrences)
